sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosTracerTest.java

```diff
@@ -0,0 +1,375 @@
+        Mockito.verify(tracerProvider, Mockito.times(4)).startSpan(Matchers.anyString(), Matchers.anyString(),
+            Matchers.anyString(), Matchers.any(Context.class));
+        verifyTracerAttributes(mockTracer, "readAllContainers." + cosmosAsyncDatabase.getId(), context,
+            cosmosAsyncDatabase.getId(), 4);
+    }
+
+    @Test(groups = {"emulator"}, timeOut = TIMEOUT)
+    public void cosmosAsyncContainer() {
+        Tracer mockTracer = Mockito.mock(Tracer.class);
+        TracerProvider tracerProvider = Mockito.spy(new TracerProvider(getMockTracer(mockTracer)));
+        ReflectionUtils.setTracerProvider(client, tracerProvider);
+        TracerProviderCapture tracerProviderCapture = new TracerProviderCapture();
+        Mockito.doAnswer(tracerProviderCapture).when(tracerProvider).startSpan(Matchers.anyString(),
+            Matchers.anyString(),
+            Matchers.anyString(), Matchers.any(Context.class));
+
+        cosmosAsyncContainer.read().block();
+        Context context = tracerProviderCapture.getResult();
+        Mockito.verify(tracerProvider, Mockito.times(1)).startSpan(Matchers.anyString(), Matchers.anyString(),
+            Matchers.anyString(), Matchers.any(Context.class));
+        verifyTracerAttributes(mockTracer, "readContainer." + cosmosAsyncContainer.getId(), context,
+            cosmosAsyncDatabase.getId(), 1);
+
+        try {
+            cosmosAsyncContainer.readThroughput().block();
+        } catch (CosmosException ex) {
+            //do nothing
+        }
+        Mockito.verify(tracerProvider, Mockito.times(2)).startSpan(Matchers.anyString(), Matchers.anyString(),
+            Matchers.anyString(), Matchers.any(Context.class));
+        verifyTracerAttributes(mockTracer, "readThroughput." + cosmosAsyncContainer.getId(), context,
+            cosmosAsyncDatabase.getId(), 2);
+
+        InternalObjectNode item = new InternalObjectNode();
+        item.setId(ITEM_ID);
+        cosmosAsyncContainer.createItem(item).block();
+        Mockito.verify(tracerProvider, Mockito.times(3)).startSpan(Matchers.anyString(), Matchers.anyString(),
+            Matchers.anyString(), Matchers.any(Context.class));
+        verifyTracerAttributes(mockTracer, "createItem." + cosmosAsyncContainer.getId(), context,
+            cosmosAsyncDatabase.getId(), 3);
+
+        cosmosAsyncContainer.upsertItem(item,
+            new CosmosItemRequestOptions()).block();
+        Mockito.verify(tracerProvider, Mockito.times(4)).startSpan(Matchers.anyString(), Matchers.anyString(),
+            Matchers.anyString(), Matchers.any(Context.class));
+        verifyTracerAttributes(mockTracer, "upsertItem." + cosmosAsyncContainer.getId(), context,
+            cosmosAsyncDatabase.getId(), 4);
+
+        cosmosAsyncContainer.readItem(ITEM_ID, PartitionKey.NONE,
+            InternalObjectNode.class).block();
+        Mockito.verify(tracerProvider, Mockito.times(5)).startSpan(Matchers.anyString(), Matchers.anyString(),
```

Comment:
Does this mean we are expecting, there would be 5 spans for a single `readItem`?

Reply:
No there will be only 1 span for readItem. This is to verify how many times this method is called in test function.

Reply:
Then that is an incorrect representation of the use case. As a testing, the test should ideally be expecting 1 span for a single API?

Reply:
We are checking incremental invocation based on api calls in single test case , which is clubbed together as per our object model , like all container apis are under one test